### DISCUSSION TITLE
Did you ever consider doing something with d3d9ex?

### DISCUSSION THREAD

**Discussion by mirh** [elishacloud/dxwrapper#105] [Ideas]
Not necessarily in relation to dxwrapper (since "fixing broken games" seems a bit different from "actively enhancing them"), but more in general.

**Comment by elishacloud** [comment]
For the most part, I don't focus much effort on enhancing the games themselves, other than the [Silent Hill 2](https://github.com/elishacloud/Silent-Hill-2-Enhancements) project.  However, there are a few minor enhancements I added to dxwrapper that may not be immediately obvious.  For example, I added [auto-de-interlacing](https://github.com/elishacloud/dxwrapper/blob/d3be2c3ab7725b71039afd9a0e0c885a772edb9f/ddraw/IDirectDrawSurfaceX.cpp#L447-L448) for some games, such as Abe's Odyssey, and [scaling options](https://github.com/elishacloud/dxwrapper/blob/d3be2c3ab7725b71039afd9a0e0c885a772edb9f/Settings/Settings.ini#L47-L54), etc.

**Comment by elishacloud** [comment]
As far as using Direct3D9Ex, I looked into this a bit in the past because it has some features, such as [overlays](https://docs.microsoft.com/en-us/windows/win32/direct3d9/d3dswapeffect#D3DSWAPEFFECT_OVERLAY) and [WaitForVBlank](https://docs.microsoft.com/en-us/windows/win32/api/d3d9/nf-d3d9-idirect3ddevice9ex-waitforvblank) that are not available on the standard Direct3D9.  However, the Ex interfaces are not as well supported so I ultimately decided not to use it.

_Edit, though I may try and do what ReShade does [here](https://github.com/crosire/reshade/blob/d6dc909f0ee11e80a5c499a80c095b410632f846/source/d3d9/d3d9.cpp#L165-L167)._

**Reply by mirh** [reply]
I'm not really sure why *Ex* would be any more or less supported than plain d3d9.. The "main underlying DDI" is still the same after all, isn't it? Except WDDM allows some extra juice on top of it. 

On the other hand, I was just thinking master @CookiePLMonster once said how *neat* upgrading games to *Ex* would be. 
I don't remember if it was about performance (even though it seemed to be for [cs:go](https://twitter.com/csgo/status/1103444404814012416) for example) or [surface sharing](https://web.archive.org/web/20120415143116/http://msdn.microsoft.com/en-us/library/windows/desktop/bb219800(v=vs.85).aspx), anyhow a legit bump in functionality if not any for interested modders to "chime in" in their turn. 
But a complete chore of an upgrade to design in the first place, so hardly worth the effort. 

But since "tinkering with apis" seems to be your working environment, maybe you had some whims about.

**Reply by elishacloud** [reply]
> I'm not really sure why _Ex_ would be any more or less supported than plain d3d9.. The "main underlying DDI" is still the same after all, isn't it? Except WDDM allows some extra juice on top of it.

I was thinking more about other wrappers.  I have encountered several Direct3D9 wrappers that have not implemented the extended interfaces.

**Comment by elishacloud** [comment]
There is an interesting thread about this here: [Optional Direct3D9Ex mode](https://github.com/crosire/d3d8to9/issues/9)

**Reply by mirh** [reply]
Lol, also with silent. 
As I hinted above *I reckon* he later amended[?](http://enbseries.enbdev.com/forum/viewtopic.php?f=2&t=6946&start=20) that point of view.

**Comment by mirh** [comment]
Seems like [dxup](https://github.com/Joshua-Ashton/dxup) also did something in this regard.

**Comment by elishacloud** [comment]
dxup is a project to convert DirectX to Vulkan.  They need to support d3d9ex because there are a handful of games that use d3d9ex and they need to be able to convert these games into Vulkan.  Converting from d3d9ex to Vulkan is different than converting from d3d7/8 to d3d9ex.

**Reply by mirh** [reply]
My point was actually more about translating d3d9 to d3d9ex (here or elsewhere)
But I see how if you are going to great lengths to upgrade the api to dx11, then maybe the switch to ex comes as a kind of a free meal. 
EDIT: btw dxup is not dxvk

**Reply by mirh** [reply]
Lol, my bad
https://steamcommunity.com/groups/SpecialK_Mods/discussions/0/215439774855607641/?ctp=23#c154645299384735387
It seems like you are gonna have a better time force upgrading to dx11, than to d3d9ex.

**Comment by mirh** [comment]
https://github.com/ersh1/Oblivion-Display-Tweaks/blob/master/oblivion_display_tweaks/D3D9Hook.cpp
https://github.com/SpecialKO/SpecialK/blob/D3D9-IK/src/render/d3d9/d3d9.cpp
Well, it seems like that there are actually tools already doing this (some on the cheapest side of things, some on the most complete)

**Comment by elishacloud** [comment]
No plans to convert d3d9 to d3d9ex.  I doubt I will ever even attempt it.  I played around with it once and to do it right would be a lot of effort and I just don't see the returns.  Also, a lot of tools don't support d3d9ex so you would lose access to those.

**Reply by mirh** [reply]
Counterstrike (and most source games IIRC?) supported it since the [longest](https://gaming.stackexchange.com/a/401019/197706) time tbh.
Though admittedly with W11 re-enabling FSO for everything anyway, most of reasons are kinda secondary (maybe there could be decreased memory consumption still on the table).
FWIW SK's implementation (which isn't perfect, but it was never exactly polished out) should be.. below 500 lines I think?

**Reply by elishacloud** [reply]
It is quite easy to convert from d3d9 to d3d9ex for a couple games, as long as the games don't use some function in d3d9 that is no longer supported in d3d9ex.  The issue would be to try and build a wrapper to convert d3d9 to d3d9ex for all games.

**Reply by mirh** [reply]
I mean.. *Something* is still better than nothing (and it would be just so much \*fun\* to see dd7to9 to 9toEx).
Though admittedly until you find a way to fix those W10 performance issues I reported, even those few potential advantages seem pointless when there's already dxvk anyway (if not any assuming remix-bridge could already relieve some of the OOM troubles when playing late 32-bit 7th generation games at higher modern resolutions).

**Comment by elishacloud** [comment]
Ok, I made a very basic conversion (https://github.com/elishacloud/dxwrapper/commit/4ebf62034b6cb900b2c8f81b18dce4fdf574e51c) from Direct3D9 to Direct3D9Ex.  The hidden option is called `D3d9to9Ex`.

This option works with `Dd7to9`, `D3d8to9` and `EnableD3d9Wrapper`.

However, Direct3D9Ex has a lot of limitations, especially with supported texture/surface formats.  Paletted textures are not supported.  16bit textures only work with system memory, etc.  I only found one game that worked and it was a 2D game.  No 3D games worked that I tested.

To get this to work right I would have to build an emulation layer to create the textures/surfaces in system memory and then convert them to 32bit textures/surfaces in video memory.  I don't know any way to do this quickly (meaning with high performance) and when I hacked a solution into place the games ran very slow.

But, anyways the feature is there and may work with newer Direct3D9 games that only use supported formats.

**Reply by elishacloud** [reply]
> I don't know any way to do this quickly (meaning with high performance) and when I hacked a solution into place the games ran very slow.

I guess I could just emulate only the Lock()/Unlock() functions.  That would be fairly fast and should work for 70% of the cases.  It might get a few more games running.  To do this right would take me at least a week or two and I just don't have time to do it now.

**Reply by mirh** [reply]
Well, damn, I wasn't expecting it.
And I could get the first 3 samples [here](https://web.archive.org/web/20160525052531/http://www.codesampler.com/dx9src/dx9src_1.htm#dx9_swap_chains) to work with it (later ones complain about D3DXCreateFont calls failing, probably due to GetDC).
FWIW I seem to understand SK overrides all `D3DPOOL_MANAGED`, to `D3DPOOL_DEFAULT` plus added `D3DUSAGE_DYNAMIC`?
By all means take your time with much more important stuff.

**Reply by elishacloud** [reply]
> I seem to understand SK overrides all `D3DPOOL_MANAGED`, to `D3DPOOL_DEFAULT` plus added `D3DUSAGE_DYNAMIC`?

I quickly made this change and it seemed to open up a lot more games.  For ddraw games, if you use `D3d9to9Ex` along side `DdrawEmulateSurface` it opens up even more games.

**Reply by elishacloud** [reply]
The build here should have much better support for D3d9to9Ex:  https://github.com/elishacloud/dxwrapper/actions/runs/15383673968

Out of the 8 or so games I tried only 1 didn't work.

**Reply by mirh** [reply]
The other samples above and [here](https://github.com/mirh/HDR_FP16x2) are still failing at device creation time, but cool I guess.

**Reply by elishacloud** [reply]
Yeah, it doesn't work well with D3DX functions.  But works pretty well with a lot of games.  I can look into this later but not high priority for me.
